interface/src/app/node/node.js

```diff
@@ -299,4 +307,123 @@ angular.module('adage.node', [
+        // is an array of the genes that this geneset contains.
+        var genesetGenes = {};
+
+        // We want to fill out this array with the genesets that score a
+        // p-value of less than the cutoff in the hypergeometric test.
+        // Each geneset in this array will be an object containing just the
+        // desired information.
+        var relevantGenesetArray = [];
+
+        // This is the main function that calculates the geneset enrichments.
+        // It calculates the enrichment for each geneset that has genes
+        // also present in the node high weight genes, and pushes that geneset
+        // into the releventGenesetArray.
+        var calculateEnrichments = function(geneGenesets, allGenesetInfo,
+                                            totalGeneNum, cutoff) {
+          var N = totalGeneNum;
+
+          // This will be the number of genes from the high weight gene list
+          // that are also present in any of the genesets that were returned
+          var m = 0;
+
+          // Fill out the genesetGenes object
+          for (var i = 0; i < $scope.hWGenes.length; i++) {
+            var genesetList = null;
+            var geneEntrezID = $scope.hWGenes[i].entrezID;
+
+            if (geneGenesets.hasOwnProperty(geneEntrezID)) {
+              genesetList = geneGenesets[geneEntrezID];
+
+              if (genesetList && genesetList.length > 0) {
+                m += 1;
+              }
+
+              for (var j = 0; j < genesetList.length; j++) {
+                var genesetID = genesetList[j];
+                if (!genesetGenes[genesetID]) {
+                  genesetGenes[genesetID] = [];
+                }
+                genesetGenes[genesetID].push($scope.hWGenes[i]);
+              }
+            } else {
+              $log.warn('Entrez ID: ' + geneEntrezID + ' not found in ' +
+                        'genesets.');
+            }
+          }
+
+          var enrichedGenesetIDs = Object.keys(genesetGenes);
+
+          for (i = 0; i < enrichedGenesetIDs.length; i++) {
+            var gsID = enrichedGenesetIDs[i];
+            var genesetInfoObj = allGenesetInfo[gsID];
+
+            var k = genesetGenes[gsID].length;
+            var n = genesetInfoObj.size;
+
+            var pValue =
+                1 - MathFuncts.hyperGeometricTest(k, m, n, N);
+            pValue = pValue.toPrecision(pValueSigDigits);
+
+            if (pValue < pValueCutoff) {
+              relevantGenesetArray.push(
+                {'name': genesetInfoObj.name, 'dbase': genesetInfoObj.dbase,
+                 'url': genesetInfoObj.url, 'pValue': pValue,
+                 'genes': genesetGenes[gsID]}
+              );
+            }
+          }
+
+          relevantGenesetArray.sort(function(a, b) {
+            return a.pValue - b.pValue;
+          });
+
+          return relevantGenesetArray;
+        };
+
+        $scope.$watch('hWGenes', function() {
```

Comment:
Since you are watching **hWGenes** by reference here, I have a feeling that it is not doing what you expected it to do. Will the interface still okay if you just write:
```
if ($scope.hWGenes.length > 0) { 
  // ... 
}
```
without the outside `$watch()` at all?